runner_service/configuration.py

```diff
@@ -141,3 +149,19 @@ def _apply_overrides(self):
             self._apply_local()
 
         self._apply_runtime()
+
+    def init_artifacts_remove(self):
+        t = threading.Thread(target=self.artifacts_remove)
+        t.start()
+
+    def artifacts_remove(self):
+        artifacts_dir = os.path.join(self.playbooks_root_dir, "artifacts")
+        dir_list = os.listdir(artifacts_dir)
+        time_now = time.mktime(time.localtime())
+        for artifacts in dir_list:
+            date = os.path.getmtime("{}/{}".format(artifacts_dir, artifacts))
+            time_difference = datetime.timedelta(seconds=time_now - date)
+            if time_difference.days >= self.artifacts_remove_age:
+                shutil.rmtree(os.path.join(artifacts_dir, artifacts))
+        time.sleep(self.artifacts_remove_age * 60 * 60 * 24)
```

Comment:
There are a some issues with this approach. First of all, as mentioned in the previous comment, you never `t.join()` the created thread, so you'll not be able to shut the application down gracefully. When you add the `t.join()` you'll end up with a different problem - consider this simple script:

```python
import threading
import time

def sleep():
    time.sleep(20)

t = threading.Thread(target=sleep)
t.start()
t.join()
```

This script will take 20 secs to execute because there's no way to interrupt the sleeping thread - we'll simply have to wait until the sleep time is over.

The algorithm you use for the cleanup also doesn't work quite as expected. Consider this scenario:
- you run your cleanup procedure as usual and schedule the next one to happen in 30 days
- one hour passes and an artifact is created
- 29 days and 23 hours pass and the cleanup procedure is running again
- you look at the artifact that was created and see that it's 29 days and 23 hours old, so < 30 days - you leave it alone
- you schedule the next cleanup run in 30 days

The effect is that artifacts could be almost as old as `2 * self.artifacts_remove_age` because the granularity is too coarse.

I think something like this should work for us:

```python
import sched
import threading
import time

def cleanup(scheduler, time_delta):
    print("cleaning...")
    # and reschedule next self-execution:
    scheduler.enter(time_delta, 0, cleanup, (scheduler, time_delta))

def cleanup_thread(running, time_delta):
    scheduler = sched.scheduler()
    # schedule first execution immediately
    scheduler.enter(0, 0, cleanup, (scheduler, time_delta))
    # and start the loop with 1 sec granularity, so join will take around 1 sec
    while running.is_set():
        scheduler.run(blocking=False)
        time.sleep(1)

running = threading.Event()
running.set()

t = threading.Thread(target=cleanup_thread, args=(running, 5))
t.start()

try:
    while True:
        print("main thread sleeping...")
        time.sleep(1)
finally:
    # application is shutting down, so let's break the loop and join
    running.clear()
    t.join()
```

Reply:
Actually my algorithm also sucks - it's too fine with the 1-sec loop. Switching from `running` event to `cancel` event makes things nicer - it allows you to have arbitrary check interval (let's say 1 hour) with immediate thread termination.

```python
import sched
import threading
import time

def cleanup(scheduler, cleanup_period):
    print("cleaning stuff older than {} secs".format(cleanup_period))
    # and reschedule next self-execution:
    scheduler.enter(cleanup_period, 0, cleanup, (scheduler, cleanup_period))

def cleanup_thread(running, check_period, cleanup_period):
    scheduler = sched.scheduler()
    # schedule first execution immediately
    scheduler.enter(0, 0, cleanup, (scheduler, cleanup_period))
    scheduler.run(blocking=False)
    # wait max check_period secs for next execution, but break immediately if necessary
    while not cancel.wait(check_period):
        print("checking if cleanup needed...")
        scheduler.run(blocking=False)

cancel = threading.Event()
cancel.clear()

# let's check in 5 secs intervals
check_period = 5

# cleanup artifacts older than 10 secs
cleanup_period = 10

t = threading.Thread(
    target=cleanup_thread,
    args=(cancel, check_period, cleanup_period)
)
t.start()

try:
    while True:
        print("main thread sleeping...")
        time.sleep(1)
finally:
    # application is shutting down, so let's break the loop and join
    cancel.set()
    t.join()
```